### DISCUSSION TITLE
Unable to specify file containing - in remap of inputs

### DISCUSSION THREAD

**Discussion by mwhapples** [hydraulic-software/conveyor#102] [Q&A]
I am trying to package an application. I start with a directory containing all the files needed. There are some platform specific subdirectories (eg. windows-x86_64, linux-x86_64, etc) which I only want distributed for the specific platform. So my thought is to add the directory to inputs and use remap to exclude the platform specific ones and then have these platform specific directories added as inputs for those machines.
My problem though is specifying the remap patterns to exclude these directories. Taking the windows-x86_64 directory as an example. If I give the pattern "-windows-x86_64/**" in the remap then windows-x86_64 is still being included. The closest I can get is giving the pattern "-windows?x86_64/**" which successfully excludes the directory.
Whilst using a wildcard to match the - works, it feels wrong as I cannot specifically match a path, the wildcard may wrongly pick up other matches.
So how should I match a - in a remap pattern?

**Comment by mikehearn** [comment]
That sounds like a bug. We will investigate!

**Reply by mwhapples** [reply]
Interestingly doing more testing of this, if the exclude is a file rather than a directory glob pattern (eg. "-some-file.txt") then that works. So I am not sure whether it relates to a glob pattern or it being part of a directory name.

**Reply by mikehearn** [reply]
I think this is a problem with the documentation. `-foo-bar/**` should work. The rules aren't actually prefix matches. The docs don't make this entirely clear. It says:

> Remap rules are tested against files not directories. Therefore, specifying a directory as a source won't work, you must use globs to do that, i.e. lib won't copy anything out of the lib directory, but lib/** will.

Does adding the glob fix it?

**Reply by mikehearn** [reply]
Or rather, this is a problem with both the documentation and the semantics. A rule that ends in a `/` is pretty clear about what the user intends, so it's kind of nonsensical to match that against a file name and then conclude it doesn't match just because it doesn't end with `**`. We'll look at altering the semantics here; I think it can be done in a backwards compatible way.

**Reply by mwhapples** [reply]
Sorry, I either got caught by markdown syntax or made a typo in my original post. I did use the glob pattern with two trailing stars and that didn't work. The pattern was meant to read "-windows-x86_64/**"

**Reply by mikehearn** [reply]
Ah yes, I see that part of your post is accidentally bolded :) 

Could you show me the entire snippet of config for the inputs you're using? Maybe it's an ordering issue.

**Reply by mwhapples** [reply]
Now I am confused, I cannot reproduce the issue as I previously found it. I had been experimenting with other patterns and now when I change it back to what I thought I had it all seems to work fine. So what is different now I don't know but where I have got to seems to work as expected. This is what is working for me:
```
    inputs += {
        from = dist
        remap = [
            "**"
            "-lib/linux-x86_64/**"
            "-lib/mac-x86_64/**"
            "-lib/windows-x86_64/**"
        ]
    }
```
Thanks for your time and sorry it turns out I cannot actually find what previously happened to give something unexpected.

**Reply by mwhapples** [reply]
I think the docs may be fine. I initially made the mistake in not noticing the need for globs, but I quickly spotted that and had accounted for it in the conf file before posting here.

**Reply by mwhapples** [reply]
Let me know if you want me to try reproducing it with a minimal example and trying to get some steps to reproduce. Otherwise I am happy with how it is working now and could easily write it off as some typo or other mistake and not spotting the error due to seeing the code I expected to see rather than the mistake.

**Reply by mikehearn** [reply]
Hmm very odd. No problem though. If you figure out what was different, let us know. Note that the logs directory contains logs for the past 20 or so runs, I think.
